lightningd/watchtower.c

```diff
@@ -0,0 +1,127 @@
+	struct pubkey final_key;
+	u8 **witness;
+	u32 remote_to_self_delay = channel->channel_info.their_config.to_self_delay;
+	const struct amount_sat dust_limit = channel->our_config.dust_limit;
+	u32 feerate_per_kw = try_get_feerate(ld->topology, FEERATE_PENALTY);
+	BUILD_ASSERT(sizeof(struct secret) == sizeof(*revocation_preimage));
+	const struct secret remote_per_commitment_secret = *revocation_preimage;
+	struct pubkey remote_per_commitment_point;
+	struct basepoints basepoints[NUM_SIDES];
+	basepoints[LOCAL] = channel->local_basepoints;
+	basepoints[REMOTE] = channel->channel_info.theirbase;
+
+	if (to_them_outnum == -1 ||
+	    amount_sat_less_eq(to_them_sats, dust_limit)) {
+		log_unusual(channel->log,
+			    "Cannot create penalty transaction because there "
+			    "is no non-dust to_them output in the commitment.");
+		return NULL;
+	}
+
+	if (!pubkey_from_secret(&remote_per_commitment_secret, &remote_per_commitment_point))
+		fatal("Failed derive from per_commitment_secret %s",
+		      type_to_string(tmpctx, struct secret,
+				     &remote_per_commitment_secret));
+
+	if (!bip32_pubkey(ld->wallet->bip32_base, &final_key,
+			  channel->final_key_idx)) {
+		fatal("Could not derive onchain key %" PRIu64,
+		      channel->final_key_idx);
+	}
+
+	if (!derive_keyset(&remote_per_commitment_point,
+			   &basepoints[REMOTE],
+			   &basepoints[LOCAL],
+			   option_static_remotekey,
+			   &keyset))
+		abort(); /* TODO(cdecker) Handle a bit more gracefully */
+	wscript = bitcoin_wscript_to_local(tmpctx, remote_to_self_delay,
+					   &keyset.self_revocation_key,
+					   &keyset.self_delayed_payment_key);
+
+	tx = bitcoin_tx(ctx, chainparams, 1, 1, locktime);
+	bitcoin_tx_add_input(tx, commitment_txid, to_them_outnum, 0xFFFFFFFF,
+			     to_them_sats, NULL);
+
+	bitcoin_tx_add_output(tx, scriptpubkey_p2wpkh(tx, &final_key),
+			      to_them_sats);
+
+	/* Worst-case sig is 73 bytes */
+	weight = bitcoin_tx_weight(tx) + 1 + 3 + 73 + 0 + tal_count(wscript);
+	weight = elements_add_overhead(weight, 1, 1);
+	fee = amount_tx_fee(feerate_per_kw, weight);
+
+	/* Result is trivial?  Spend with small feerate, but don't wait
+	 * around for it as it might not confirm. */
+	if (!amount_sat_add(&min_out, dust_limit, fee))
+		log_broken(channel->log,
+			      "Cannot add dust_limit %s and fee %s",
+			      type_to_string(tmpctx, struct amount_sat, &dust_limit),
+			      type_to_string(tmpctx, struct amount_sat, &fee));
+
+	if (amount_sat_less(to_them_sats, min_out)) {
+		/* FIXME: We should use SIGHASH_NONE so others can take it */
+		fee = amount_tx_fee(feerate_floor(), weight);
+	}
+
+	/* This can only happen if feerate_floor() is still too high; shouldn't
+	 * happen! */
+	if (!amount_sat_sub(&amt, to_them_sats, fee)) {
+		amt = dust_limit;
+		log_broken(channel->log,
+			   "TX can't afford minimal feerate"
+			   "; setting output to %s",
+			   type_to_string(tmpctx, struct amount_sat, &amt));
+	}
```

Comment:
in the case that the penalty output is dust, is it still worth issuing a penalty tx? I'm guessing that the watchtower needs to be updated to the new state *every* commitment so it still needs something.

Reply:
I was also wondering about this. I'm trying it atm and realised that until the dust limit is not passed, no penalty is created. 

```
2020-04-07T11:01:08.782Z DEBUG 020e188b565427822990c7c599f772f68e095d7f93dfacaaacfc6c8239ad37ef08-chan#2: Balance 500000000msat -> 499999948msat
2020-04-07T11:01:08.783Z UNUSUAL 020e188b565427822990c7c599f772f68e095d7f93dfacaaacfc6c8239ad37ef08-chan#2: Cannot create penalty transaction because there is no non-dust to_them output in the commitment.
```

The commitments are update anyway aren't they? If so this could be a problem since the tower would not be covering some cases.

Reply:
This should only be an issue until the fundee goes above the reserve value, before then it makes little sense to create a penalty transaction at all, since the counterparty doesn't have funds to lose in the first place. The reserve value guarantees that there is at least one non-dust outputs once both parties have surpassed the reserve value:

[BOLT 02](https://github.com/lightningnetwork/lightning-rfc/blob/f068dd0d8dfa5ae75feedd99f269e23be4777381/02-peer-protocol.md#requirements):
> MUST set channel_reserve_satoshis greater than or equal to dust_limit_satoshis.

So we only need to worry about `reserve_value < dust + minfeerate * penaltytx_weight`. In which case we'd still get the majority of funds (otherwise the penalty would actually be worth it) and we're ok to give our peer a couple of breadcrumbs that would have been his in that state anyway. The loss therefore is limited to `minfeerate * penatlytx_weight` :-)

But it also means that there may be state updates that intermittently do not result in a penalty_tx being created, which as far as I can see is part of the protocol and would happen in all implementations.